### PR TITLE
Refactor AppDelegate into a trait

### PR DESCRIPTION
**Pull request by LiHRaM**
While working on #268, @cmyr and I decided it might be worthwhile to implement `AppDelegate` as a trait rather than an object storing closures.
This is the initial attempt, and it feels nice and simple.

I implemented the trait for an empty struct in `multiwin`, which smells a bit.
This approach makes it possible for the AppDelegate to maintain state separately from the data which is passed to it via events, any thoughts on how that could be useful? Otherwise this approach might need some tweaking.

### CODE REVIEW

**Comment by cmyr** [comment]
An impl for an empty struct is totally reasonable.

The 'should the app delegate have secret state' question is a deep one. I really don't know; the main justification for the previous design was to make this impossible, but I didn't have a good justification for wanting this to be impossible. 

More and more, lately, one of my main design principles for druid is that we should make it easy for framework users to implement the hacks they need to get their apps working, and a stateful app delegate is another tool in that drawer.

File: druid/examples/multiwin.rs [hunk 1]
```diff
@@ -81,8 +82,16 @@ fn ui_builder() -> impl Widget<State> {
     col
 }
 
-fn make_delegate() -> AppDelegate<State> {
-    AppDelegate::new().event_handler(|event, data, _env, ctx| {
+struct MultiwinDelegate {}
```
Comment (cmyr) [inline druid/examples/multiwin.rs]: For a struct with no fields I prefer the 'unit struct' syntax, e.g. you can just write,

```rust
struct Delegate;
```

File: druid/src/app.rs [hunk 2]
```diff
@@ -70,7 +70,7 @@ impl<T: Data + 'static> AppLauncher<T> {
     /// Set the [`AppDelegate`].
     ///
     /// [`AppDelegate`]: struct.AppDelegate.html
-    pub fn delegate(mut self, delegate: AppDelegate<T>) -> Self {
+    pub fn delegate(mut self, delegate: Box<dyn AppDelegate<T>>) -> Self {
```
Comment (cmyr) [inline druid/src/app.rs]: I would make this argument be, `impl AppDelegate<T>`, and then we can box inside the function; this just saves the caller from the hassle of having to box before they call.

(you might also need a `+ 'static` after the type)

File: druid/src/app_delegate.rs [hunk 3]
```diff
@@ -57,46 +55,18 @@ type EventFn<T> = dyn Fn(Event, &mut T, &Env, &mut DelegateCtx) -> Option<Event>
...
-            None => Some(event),
-        }
-    }
-}
-
-impl<T: Data> Default for AppDelegate<T> {
-    fn default() -> Self {
-        AppDelegate::new()
-    }
+    ) -> Option<Event>;
```
Comment (cmyr) [inline druid/src/app_delegate.rs]: I would give this a default impl, which which is just `Some(event)`, e.g. it just passes through the event unchanged.

My thinking is that, on the app delegate, all methods are optional.

Comment (LiHRaM) [reply druid/src/app_delegate.rs]: Ah, yes, I had this thought too, but I've just iterated through several redesigns and totally forgot it. I'll add that one for sure. :+1:

Comment (LiHRaM) [reply druid/src/app_delegate.rs]: So the compiler complains about unused variables, but I don't want to underscore the names, as they'll be part of the docs. Ergo the current approach, as suggested in https://github.com/rust-lang/rust/issues/26487.

File: druid/src/app_delegate.rs [hunk 4]
```diff
@@ -57,46 +55,23 @@ type EventFn<T> = dyn Fn(Event, &mut T, &Env, &mut DelegateCtx) -> Option<Event>
...
-        }
-    }
-}
-
-impl<T: Data> Default for AppDelegate<T> {
-    fn default() -> Self {
-        AppDelegate::new()
+        let _ = data;
+        let _ = env;
+        let _ = ctx;
```
Comment (cmyr) [inline druid/src/app_delegate.rs]: what's this about?

Comment (cmyr) [reply druid/src/app_delegate.rs]: oh, unused warnings? I would annotate this function `#[allow(unused)]`.

Comment (LiHRaM) [reply druid/src/app_delegate.rs]: Yeah, I was following the discussion mentioned in https://github.com/xi-editor/druid/pull/283#discussion_r344462460, and it seemed to be the most correct solution, but the attribute is also fine with me. 😄

Comment (LiHRaM) [reply druid/src/app_delegate.rs]: Done!

Comment (cmyr) [reply druid/src/app_delegate.rs]: ah gotcha, missed that comment. We have adopted the pattern of using `#[allow]` on each function with this problem, so let's just keep doing that. 🤷‍♂

**Comment by cmyr** [comment]
feel free to merge. :)

**Pull request merged**